Automattic/woocommerce-payments
# Allow users with a migrated WCPay Subscription to pay for a failed renewal order

Fixes #6812 

#### Changes proposed in this Pull Request

WCPay Subscriptions that have failed renewal orders go through a very specific flow to update the payment method and reattempt the payment. 

To achieve this it alters the 'Pay' order action URL to redirect the customer to the subscription update payment method flow instead of the default order pay flow. 

If a customer also manages to land on pay for order page where the order belongs to a WCPay subscription that needs a payment method update, it redirects the customer to the correct flow. 

This logic has a bug in it though that prevents a customer from paying for a failed renewal order that was generated by a WCPay subscription, if that subscription is subsequently migrated to a tokenized subscription.

This PR consolidates the logic used to determine if an order Pay URL needs to change or the redirect needs to happen. This function ensures it also takes into consideration if the subscription has been migrated. 

#### Testing instructions

1. Purchase a subscription using WCPay subscriptions. 
2. Using the billing clocks process a failed renewal order. 
3. Migrate the subscription to a tokenized subscription by deleting the `_wcpay_subscription_id` subscription meta.
4. Install and activate WC Subscriptions.
5. Log in to the billing clock user and attempt to pay for the order. 
   - On `develop` you will get the following error notice.
   - On this branch you shouldn't see any error. 

![image](https://github.com/Automattic/woocommerce-payments/assets/8490476/7f11a988-ea18-4067-ad79-885daf94d03e)

1. Deactivate WC Subscriptions.
2. Repeat steps 1-2 from above. 
3. Log in to the billing clock user and attempt to pay for the order. 
   - On this branch you shouldn't see any issue you should be redirected to the update payment method flow. As expected. 

-------------------


**Post merge**
